### DISCUSSION TITLE
How to implement next and previous post links

### DISCUSSION THREAD

**Discussion by CharliePoole** [statiqdev/Discussions#24] [Q&A]
I'd like to modify `_posts.cshtml` so that each post (IsPost: true) displays links for the next and previous post.

Existing `Next` and `Previous` properties of the document relate to pagination, so they are no help and are null anyway. Further, the ideal solution would give a different value for next and previous according to the particular archive from which we arrived at the post. So, if we came from the main Index page, we'd get the next pst on that list, if we came from a tag we'd get the next item from that archive.

Thinking about this, I've come to the conclusion that I need to have some sort of properties set on the originating list itself. But before digging into  it I'd like to make sure there isn't something existing I can hook into.

If I need to solve it myself, it seems to me that I need to deal with two distinct problems:

1. The post must "know" what archive it was linked from. If not from an archive, I can default to the main Index or not display any links at all.

2. The archive document has to provide some property that will give the next and previous docs.

All said, my preference is to use what's there already.

**Comment by daveaglick** [comment]
That makes perfect sense. There might already be some support here to help you...

The challenge with next/previous traversal is that the documents can be cloned along the way (and often are). So at whatever stage you "lock in" the next/previous metadata, if the set of documents are cloned after that, the `Next` metadata would still reference the original pre-cloned document. That creates a ton of confusion because you end up with next/previous values that seem like they're correct but the document they point to is missing some metadata you think it should have, or doesn't have rendered content, etc.

To address this, I created extensions on `IEnumerable<T>` to get things like next/previous. The trick is to call the `IEnumerable<IDocument>.GetNext(currentDocument)` from the right  collection. For example, the following adds next and previous buttons to every post from inside the `_post-footer.cshtml` file (just for illustration though since the order isn't well sorted here, though I guess you could sort `Inputs` before getting next/previous):

```
@{
    IDocument older = Context.Inputs.GetPrevious(Document);
    IDocument newer = Context.Inputs.GetNext(Document);
}
@if (older != null || newer != null)
{
    <div>
        <div class="float-sm-left">
            @if (newer != null)
            {
                <a class="btn btn-primary" href="@Context.GetLink(newer)" role="button"><i class="fas fa-angle-double-left"></i> Newer</a>
            }
        </div>
        <div class="float-sm-right">
            @if (older != null)
            {
                <a class="btn btn-primary" href="@Context.GetLink(older)" role="button">Older <i class="fas fa-angle-double-right"></i></a>
            }
        </div>
        <div class="clearfix"></div>
    </div>
}
```

Wanting that to change based on archive is tricky though since the archives are all about showing lists and link to the same document regardless of where the link originated. One approach might be to calculate next/previous from each archive page and pass that in as a query string. Another would be to pass an archive name or some other identifier as a query string and then calculate all possible next/previous links during generation and toggle in JS from that identifier.

Let me noodle on this one a bit and see if I come up with anything better.

**Reply by daveaglick** [reply]
After thinking about it, I haven't really come up with anything better. It's the dynamic buttons based on context (where the request came from) that makes this tricky. If I were implementing this for my site I think I'd go with the approach of embedding forward/back links on each page for every context and then switch them from JS based on a query string. That way I could continue to flow the query string through each forward/back link and keep that context persisted.

**Comment by CharliePoole** [comment]
I'm waiting to see how badly it seems to be needed before doing anything else.
